pydra/engine/specs.py

```diff
@@ -738,43 +678,71 @@ class SingularitySpec(ContainerSpec):
     container: str = attr.ib("singularity", metadata={"help_string": "container type"})
 
 
-class LazyField:
-    """Lazy fields implement promises."""
-
-    def __init__(self, node, attr_type):
-        """Initialize a lazy field."""
-        self.name = node.name
-        if attr_type == "input":
-            self.fields = [field[0] for field in node.input_spec.fields]
-        elif attr_type == "output":
-            self.fields = node.output_names
-        else:
-            raise ValueError(f"LazyField: Unknown attr_type: {attr_type}")
-        self.attr_type = attr_type
-        self.field = None
+@attr.s
+class LazyInterface:
+    _node: "core.TaskBase" = attr.ib()
+    _attr_type: str
 
     def __getattr__(self, name):
-        if name in self.fields or name == "all_":
-            self.field = name
-            return self
-        if name in dir(self):
-            return self.__getattribute__(name)
-        raise AttributeError(
-            f"Task {self.name} has no {self.attr_type} attribute {name}"
+        if name not in self._field_names:
+            raise AttributeError(
+                f"Task {self._node.name} has no {self._attr_type} attribute {name}"
+            )
+        return LazyField(
+            name=self._node.name,
+            field=name,
+            attr_type=self._attr_type,
+            type=self._get_type(name),
         )
 
-    def __getstate__(self):
-        state = self.__dict__.copy()
-        state["name"] = self.name
-        state["fields"] = self.fields
-        state["field"] = self.field
-        return state
 
-    def __setstate__(self, state):
-        self.__dict__.update(state)
+class LazyIn(LazyInterface):
+    _attr_type = "input"
+
+    def _get_type(self, name):
+        attr = next(t for n, t in self._node.input_spec.fields if n == name)
+        if attr is None:
+            return ty.Any
+        else:
+            return attr.type
+
+    @property
+    def _field_names(self):
+        return [field[0] for field in self._node.input_spec.fields]
+
+
+class LazyOut(LazyInterface):
+    _attr_type = "output"
+
+    def _get_type(self, name):
+        try:
+            type_ = next(f[1] for f in self._node.output_spec.fields if f[0] == name)
+        except StopIteration:
+            type_ = ty.Any
+        else:
+            if not inspect.isclass(type_):
+                try:
+                    type_ = type_.type  # attrs _CountingAttribute
+                except AttributeError:
+                    pass  # typing._SpecialForm
+        return type_
+
+    @property
+    def _field_names(self):
+        return self._node.output_names + ["all_"]
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class LazyField:
+    """Lazy fields implement promises."""
+
+    name: str
+    field: str
+    attr_type: str
+    type: ty.Type[ty.Any]
 
     def __repr__(self):
-        return f"LF('{self.name}', '{self.field}')"
+        return f"LF('{self.name}', '{self.field}', {self.type})"
```

Comment:
IDK if this would be more readable? Depends on the `repr` of `self.type`. For types

```suggestion
        return f"Lazy[{self.type}]('{self.name}', '{self.field}')"
```